mitmproxy/protocol/http.py

```diff
@@ -187,8 +187,7 @@ def __call__(self):
                     flow.request.set_auth(self.config.upstream_auth)
                 self.process_request_hook(flow)
 
-                if not flow.response:
-                    self.establish_server_connection(flow)
+                if not flow.response and self.establish_server_connection(flow):
```

Comment:
Why don't we just raise a proper exception in `establish_server_connection` then?